ledger/heavy/executor/jetkeeper.go

```diff
@@ -20,143 +20,355 @@ import (
 
 	"github.com/insolar/insolar/insolar"
 	"github.com/insolar/insolar/insolar/jet"
-	"github.com/insolar/insolar/instrumentation/inslogger"
 	"github.com/insolar/insolar/internal/ledger/store"
 )
 
 //go:generate minimock -i github.com/insolar/insolar/ledger/heavy/executor.JetKeeper -o ./ -s _gen_mock.go
 
 // JetKeeper provides a method for adding jet to storage, checking pulse completion and getting access to highest synced pulse.
 type JetKeeper interface {
-	// Add performs adding jet to storage and checks pulse completion.
-	Add(context.Context, insolar.PulseNumber, insolar.JetID) error
+	// AddDropConfirmation performs adding jet to storage and checks pulse completion.
+	AddDropConfirmation(ctx context.Context, pn insolar.PulseNumber, jet insolar.JetID, split bool) error
+	// AddHotConfirmation performs adding hot confirmation to storage and checks pulse completion.
+	AddHotConfirmation(ctx context.Context, pn insolar.PulseNumber, jet insolar.JetID, split bool) error
 	// TopSyncPulse provides access to highest synced (replicated) pulse.
 	TopSyncPulse() insolar.PulseNumber
 }
 
-func NewJetKeeper(jets jet.Storage, db store.DB) JetKeeper {
-	return &dbJetKeeper{jetTrees: jets, db: db}
+func NewJetKeeper(jets jet.Storage, db store.DB, pulses pulse.Calculator) JetKeeper {
+	return &dbJetKeeper{
+		jetTrees: jets,
+		db:       db,
+		pulses:   pulses,
+	}
 }
 
 type dbJetKeeper struct {
 	jetTrees jet.Storage
 
+	pulses storage.PulseCalculator
+
 	sync.RWMutex
 	db store.DB
 }
 
-func (jk *dbJetKeeper) Add(ctx context.Context, pulse insolar.PulseNumber, id insolar.JetID) error {
+type jetInfo struct {
+	JetID         insolar.JetID
+	HotConfirmed  []insolar.JetID
+	DropConfirmed bool
+	Split         bool
+}
+
+func (j *jetInfo) addDrop(newJetID insolar.JetID, split bool) error {
+	if j.DropConfirmed {
+		return errors.New("addDrop. try to rewrite drop confirmation. existing: " + j.JetID.DebugString() +
+			", new: " + newJetID.DebugString())
+	}
+	j.Split = split
+	j.DropConfirmed = true
+	j.JetID = newJetID
+
+	return nil
+}
+
+func (j *jetInfo) checkIncomingHot(incomingJetID insolar.JetID) error {
+	if len(j.HotConfirmed) >= 2 {
+		return errors.New("num hot confirmations exceeds 2. existing: " + insolar.JetIDCollection(j.HotConfirmed).DebugString() +
+			", new: " + incomingJetID.DebugString())
+	}
+
+	if len(j.HotConfirmed) == 1 && j.HotConfirmed[0].Equal(incomingJetID) {
+		return errors.New("try add already existing hot confirmation: " + incomingJetID.DebugString())
+	}
+
+	return nil
+}
+
+func (j *jetInfo) addHot(newJetID insolar.JetID, parentId insolar.JetID, split bool) error {
+	err := j.checkIncomingHot(newJetID)
+	if err != nil {
+		return errors.Wrap(err, "incorrect incoming jet")
+	}
+
+	j.HotConfirmed = append(j.HotConfirmed, newJetID)
+	j.JetID = parentId
+
+	return nil
+}
+
+func (j *jetInfo) isConfirmed() bool {
+	if !j.DropConfirmed {
+		return false
+	}
+	if len(j.HotConfirmed) == 0 || j.HotConfirmed[0].IsEmpty() {
```

Comment:
Why `j.HotConfirmed[0].IsEmpty()`? What if the second is empty. Its probably not necessary, it wont pass the check if its empty anyway.